### DISCUSSION TITLE
How to write a new layer to the current dwg

### DISCUSSION THREAD

**Discussion by yingyue4008** [LibreDWG/libredwg#579] [Q&A]
I first read the data of dwg, and then created a new layer, but I crashed when writing dwg, so I don't know how to write the newly created layer?

**Comment by rurban** [comment]
How did you create the new layer? It should be done with dwg_add_LAYER()

**Comment by yingyue4008** [comment]
I use dwg_ add_ LAYER(), but the parameter Dwg_ Data * restrict dwg cannot be obtained from readdwg, but only through dwg_ add_ Document function add Dwg_ Data, and I want to use readdwg's Dwg_ Data

**Reply by rurban** [reply]
Yes, you need to use dwg from your readdwg

**Reply by yingyue4008** [reply]
But call dwg_ write_ File function, crash

**Reply by yingyue4008** [reply]
In fact, I just want to modify the existing line entity and add a new layer

**Reply by yingyue4008** [reply]
> Yes, you need to use dwg from your readdwg

Excuse me, it is still impossible to write the new layer into the read dwg. Is this the right way?

**Reply by rurban** [reply]
Yes

**Reply by yingyue4008** [reply]
> Yes
Then how should I use it? Hope to be prompted

**Reply by rurban** [reply]
https://github.com/LibreDWG/libredwg/blob/master/examples/load_dwg.c has such an example how to read a DWG and then add something

**Reply by yingyue4008** [reply]
> https://github.com/LibreDWG/libredwg/blob/master/examples/load_dwg.c has such an example how to read a DWG and then add something

This logic is only read, and does not add data in the original dwg drawing. Please help me see if this is the case

**Comment by rurban** [comment]
I've updated https://github.com/LibreDWG/libredwg/blob/smoke/load_dwg/examples/load_dwg.c to show how to load, check, add a text to a dwg. Similar for a layer.
examples/dwgadd.c has a command to add a layer also.

**Reply by yingyue4008** [reply]
> I've updated https://github.com/LibreDWG/libredwg/blob/smoke/load_dwg/examples/load_dwg.c to show how to load, check, add a text to a dwg. Similar for a layer. examples/dwgadd.c has a command to add a layer also.

Thank you very much. Let me learn first

**Reply by yingyue4008** [reply]
> I've updated https://github.com/LibreDWG/libredwg/blob/smoke/load_dwg/examples/load_dwg.c to show how to load, check, add a text to a dwg. Similar for a layer. examples/dwgadd.c has a command to add a layer also.

I found a problem, only call once dwg_ add_ LAYER() is no problem, but it can't be written into dwg after another call dwg_ add_ LAYER()

**Reply by rurban** [reply]
Oh, this looks like a bug then. will check, thanks

**Reply by yingyue4008** [reply]
> Oh, this looks like a bug then. will check, thanks

That's all right.I have been using the content here recently. I hope to use it as soon as possible. Thank you very much for your efforts

**Reply by yingyue4008** [reply]
> I've updated https://github.com/LibreDWG/libredwg/blob/smoke/load_dwg/examples/load_dwg.c to show how to load, check, add a text to a dwg. Similar for a layer. examples/dwgadd.c has a command to add a layer also.

Excuse me, I will replace the old layer on the line with the new layer, and then write it into dwg. Open it to see if it has not been replaced. Could you tell me if this is my operation error？

Dwg_Object_LAYER* pNLayer = dwg.object[i].tio.entity->layer->obj->tio.object->tio.LAYER;
pNLayer = pNewLayer;

**Reply by rurban** [reply]
No, this will not work. Layer handles need to be replaced in the LAYER_CONTROL list also

**Reply by yingyue4008** [reply]
> No, this will not work. Layer handles need to be replaced in the LAYER_CONTROL list also
Thank you very much. I will try this method

**Reply by yingyue4008** [reply]
> No, this will not work. Layer handles need to be replaced in the LAYER_CONTROL list also

When encountering similar problems, rewrite LAYER_  CONTROL After, perform dwg_ write_ File, the program crashed. I hope to provide an example for help

**Reply by rurban** [reply]
Maybe we'd need a dwg_delete_OBJECT() API also

**Reply by yingyue4008** [reply]
> dwg_delete_OBJECT

Do you need to delete the original layer on the line and assign a new layer?

**Reply by yingyue4008** [reply]
> No, this will not work. Layer handles need to be replaced in the LAYER_CONTROL list also

Excuse me, can you provide an example? Let me compare it to see if it is wrong

**Reply by rurban** [reply]
Only if you use dwg_add_LAYER and need to delete the old one. Normally add is enough, AutoCAD does it the same.

**Reply by yingyue4008** [reply]
> Only if you use dwg_add_LAYER and need to delete the old one. Normally add is enough, AutoCAD does it the same.

At present, I haven't achieved success after operating in the above way, so I don't know where the error occurred

**Reply by yingyue4008** [reply]
This is how I operate：
  for (iControl = 0; iControl < num_layers; iControl++)
     {
                Dwg_Object* obj = dwgRefObject(&dwg, _ctrl->entries[iControl]);
                if (obj && obj->fixedtype == DWG_TYPE_LAYER)
                {
                     if (pNLayer == obj->tio.object->tio.LAYER)
                     {
                           obj->tio.object->tio.LAYER = pNewLayer;
                     }
               }
}

**Reply by rurban** [reply]
You cannot do that. You need to replace single LAYER properties only, like the text or some flags.

**Reply by yingyue4008** [reply]
> You cannot do that. You need to replace single LAYER properties only, like the text or some flags.

Thank you very much for your reply, but I'm a newly created layer. I want to replace the layer on the original line with a new one, so I don't know what to do

**Reply by rurban** [reply]
Then you add a new layer, and use this layer handle on the LINE.

**Reply by yingyue4008** [reply]
> Then you add a new layer, and use this layer handle on the LINE.

Hello,Do you need to call a transaction after creating a new layer?

**Reply by yingyue4008** [reply]
> Then you add a new layer, and use this layer handle on the LINE.

Debugging found that the new layer ->xref  is empty. How to set xref?

**Reply by yingyue4008** [reply]
> Then you add a new layer, and use this layer handle on the LINE.
Hello, excuse me, I haven't replaced the correct layer at present. I still don't know how to use the layer handle to give the line

**Reply by yingyue4008** [reply]
Excuse me, at present, this layer replacement has not been successful. I don't know if it is the reason for my incorrect operation
